tap_zendesk/client.py

```diff
@@ -129,8 +121,135 @@ def parse_response(self, response: requests.Response) -> Iterable[dict]:
+            else None,
+        )
+
+        # Check for 404 Not Found and log a warning
+        if response.status_code == 404:
+            self.logger.warning(f"Received 404 for URL: {response.url}")
+            return None
+
+        self.validate_response(response)
+        self.logger.debug("Response received successfully.")
+        return response
+
+    def get_records(self, context):
+        end_date_str = self.config.get("end_date")
+        end_date = datetime.fromisoformat(end_date_str) if end_date_str else None
+
+        if end_date and end_date.tzinfo is None:
+            end_date = end_date.replace(tzinfo=timezone.utc)
+
+        replication_key = self.replication_key
+
+        paginator = self.get_new_paginator()
+        next_page_token = None
+
+        while True:
+            prepared_request = self.prepare_request(context, next_page_token)
+            response = self._request(prepared_request, context)
+            if not response:
+                self.logger.error("Received empty response")
+                break
+
+            records = self.parse_response(response)
+
+            for record in records:
+                if not record:
+                    self.logger.error("Received empty record")
+                    continue
+
+                try:
+                    record_json = json.loads(json.dumps(record))
+                except json.JSONDecodeError as e:
+                    self.logger.error(f"JSONDecodeError encountered while parsing record: {e}")
+                    continue
+
+                record_date_str = record.get(replication_key)
+                if record_date_str:
+                    self.logger.debug(record_date_str)
+                    record_date = datetime.fromisoformat(record_date_str)
+                    if record_date.tzinfo is None:
+                        record_date = record_date.replace(tzinfo=timezone.utc)
+                    if end_date and record_date > end_date:
+                        self.logger.info(
+                            f"Stopping data fetch as record date {record_date} exceeds end date {end_date}")
+                        return
+
+                yield record
+
+            next_page_token = self.get_next_page_token(response)
+            self.logger.debug(f"Next page token: {next_page_token}")
+            if not next_page_token or self.is_end_of_stream(response):
+                break
+
+    def get_next_page_token(self, response):
+        """Get the next page token from the response, if applicable."""
+        if 'after_cursor' in response.json():
+            next_page_token = response.json().get('after_cursor')
+            self.logger.debug(f"Next page token (cursor): {next_page_token}")
+            return next_page_token
+        next_page_token = response.json().get('meta', {}).get('after_cursor')
+        self.logger.debug(f"Next page token (meta): {next_page_token}")
+        return next_page_token
```

Comment:
We can use the `next_page_token_jsonpath` to send only the path to the next cursor an we will don't need to create this method (only if we want to check filter the end_data, but we can call the super method in this case too).